core/iwasm/libraries/thread-mgr/thread_manager.c

```diff
@@ -76,6 +76,66 @@ traverse_list(bh_list *l, list_visitor visitor, void *user_data)
     }
 }
 
+/* Assumes cluster->lock is locked */
+static bool
+safe_traverse_exec_env_list(WASMCluster *cluster, list_visitor visitor,
+                            void *user_data)
+{
+    Vector proc_nodes;
+    void *node, *next;
+    bool ret = true;
+
+    if (!bh_vector_init(&proc_nodes, cluster->exec_env_list.len, sizeof(void *),
+                        false)) {
+        ret = false;
+        goto final;
+    }
+
+    node = bh_list_first_elem(&cluster->exec_env_list);
+    cluster->exec_env_list_element_removed = false;
+
+    while (node) {
+        bool already_processed = false;
+        void *proc_node;
+        for (size_t i = 0; i < bh_vector_size(&proc_nodes); i++) {
+            if (!bh_vector_get(&proc_nodes, i, &proc_node)) {
+                ret = false;
+                goto final;
+            }
+            if (proc_node == node) {
+                already_processed = true;
+                break;
+            }
+        }
+        if (already_processed) {
+            node = bh_list_elem_next(node);
+            continue;
+        }
+
+        next = bh_list_elem_next(node);
+        os_mutex_unlock(&cluster->lock);
+        visitor(node, user_data);
+        os_mutex_lock(&cluster->lock);
+        if (!bh_vector_append(&proc_nodes, &node)) {
+            ret = false;
+            goto final;
+        }
+
+        if (cluster->exec_env_list_element_removed) {
+            cluster->exec_env_list_element_removed = false;
```

Comment:
Yeah, good point. I can think of two solutions for that:
1. have a new lock that guards the whole function (so no two functions can execute at the same time)
2. Update the function so it always assumes the element was removed

I feel the 1) is safer as the 2) might cause re-running the same visitor twice which might not be desired (it's used in scenarios where elements are removed from the list), although it adds one more field to the cluster object (maybe that's not a big issue though?). @wenyongh  what are your thoughts?

Reply:
Not sure whether 1) will cause dead lock or not? For example, thread A locks the lock for this function and then unlock cluster->lock, after handling the node, it is waiting the cluster->lock, and thread B locks cluster->lock and is waiting the lock for this function?

Reply:
I don't think that situation can happen in the current setup, but I see in the future incorrectly using the function might cause some problems. Would you say then 2) would be better approach? I think we can still keep an array of already processed nodes to avoid re-processing elements. That will be less efficient for large number of exec_env's, but I don't think the number will ever grow significantly?

Reply:
What we are afraid is that the node visited is removed by other threads so we cannot use its next again, right?
If yes, after visiting and locking cluster->lock again, can we search the whole list to check whether the node is in the list? If yes, node = node->next, else node = bh_list_first_elem(..), does this way work?

Reply:
Yeah, so what you're suggesting is a slightly improved (more performant in some cases) version of 2). I think that could work. The only concern is when two threads call the traversal at the same time, although for current usecase it's not the problem because the traversal callback does another lock and make sure the node exists. I'll update the PR with the implementation.

Reply:
@wenyongh I updated the code, I'd appreciate you  taking a look.